063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,221 @@
+    "recentWindows": 5,
+    "dataToMoveMB": 0,
+    "excludedTopics": [],
+    "numLeaderMovements": 0
+  }
+  ...
+```
+
+The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.goals` property.
+
+The `replication_throttle` parameter is an upper bound on the bandwidth used to move replicas (bytes per second).
+This is configurable for all endpoints in `Kafka.spec.cruiseControl.config` with `default.replication.throttle` property.
+
+The `skip_rack_awareness_check` boolean parameter is undocumented and disabled by default.
+In case of a single-rack deployment, rack goals are not configured and the Topic Operator will automatically enable `skip_rack_awareness_check`.
+
+#### User tasks endpoint
+
+The `user_tasks` endpoint is used to periodically check the tasks state by `User-Task-ID`, which is returned in the `topic_configuration` response.
+
+```sh
+$ curl -s "localhost:9090/kafkacruisecontrol/user_tasks?user_task_ids=5344ca89-351f-4565-8d0f-9aade00e053d&json=true" | jq
+{
+  "userTasks": [
+    {
+      "Status": "Completed",
+      "ClientIdentity": "127.0.0.1",
+      "RequestURL": "POST /kafkacruisecontrol/topic_configuration?topic=my-topic&replication_factor=2&skip_rack_awareness_check=true&dryrun=false&json=true",
+      "UserTaskId": "5344ca89-351f-4565-8d0f-9aade00e053d",
+      "StartMs": "1701961135978"
+    }
+  ],
+  "version": 1
+}
+```
+
+### Topic Operator
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic also detects `.spec.replicas` changes by comparing with the actual topic replication factor in Kafka.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- **Pending**: replicas is updated, but the replication factor change request is not sent to Cruise Control.
+- **Ongoing**: replicas is updated and Cruise Control has accepted the request, but the execution is not completed.
+- **Ready**: replicas is updated and Cruise Control execution is completed.
+
+When Cruise Control integration is enabled and a replicas change is detected by the BatchingTopicController, the event is notified to the ReplicasChangeManager.
+The ReplicasChangeManager is responsible for managing the communication with Cruise Control's REST API and updating the KafkaTopic status.
+
+<p align="center"><img src="images/063-topic-replication-factor-change.png"/></p>
+
+The KafkaTopic resources with replicas change are added to a pendingChanges list the first time they are notified to the ReplicasChangeManager.
+When a replicas change is marked as pending, other replicas changes for the same KafkaTopic are ignored until the pending replicas change is sent to Cruise Control and the execution is completed.
```

Comment:
* How would this be implemented?
* Why would this be the case if the Pending state means it is not queued in CC yet? It seems in such case you can easily accept new changes or just cancel it.

Reply:
The periodic reconciliation logic continues to detect and notify the replicas change to the ReplicasChangeManager. After the first, the ReplicasChangeManager ignores other replicas change notifications for the same topic using the pendingChanges cache. Once the current pending is done, the new change becomes pending and can be executed.

This has the benefit of ignoring intermediate changes in case they are done in a relatively short period of time, and only apply the first and last change, also considering that CC runs one task at a time, and the queue is shared with all other tasks (i.e. rebalances).

When a new replicas change for the same topic comes in, we may consider stopping the current replicas change by removing it from pendingChanges cache, and doing a stop request to CC in case it is also in ongoingChanges. I didn't do that, because it would be a lot of work and wasted cycles for CC. When you stop a task, the current batch still completes, and it could take some time.

Reply:
It seems that this would not survive restarts of the operator. Also, there seems to be no need for this. At least until the task is queued within CC (that seems to include the pending state mentioned here) there is no reason to not reflect the latest change.

Reply:
> It seems that this would not survive restarts of the operator. 

Pending changes that are also ongoing, will be restored on ReplicasChangeManager initialization when the TO restarts. The rest of pending changes will be added as if they were new as soon as the first periodic reconciliation runs.

> Also, there seems to be no need for this. At least until the task is queued within CC (that seems to include the pending state mentioned here) there is no reason to not reflect the latest change.

The change is reflected in KafkaTopics, but with a Condition that says ongoing, so I think it is correct.

Reply:
I still don't think this is recoverable. The condition does not store the number of replicas. So if the operator restarts, you cannot recover from it what the change was. I also think recovering from the conditions is weird. I haven't seen that done anywhere. I would not use conditions to store information but only to inform users.